### ISSUE TITLE
Non-JavaScript syntax in JavaScript files

### ISSUE THREAD

**Issue by Sequoia** [facebook/flow#3901]
I attempted to run the code from the following simple example **JavaScript** file ([examples/02_TypeAnnotations/type_annotations.js](https://github.com/facebook/flow/blob/master/examples/02_TypeAnnotations/type_annotations.js)) in JS Bin ([example here](http://jsbin.com/wahawawugu/edit?js,console)), a popular online tool to test and share **JavaScript** code, and it produced an error:

```
"SyntaxError: Unexpected token :
    at https://static.jsbin.com/js/prod/runner-3.41.10.min.js:1:13926
    at https://static.jsbin.com/js/prod/runner-3.41.10.min.js:1:10855"
```

I tried it in Node.js as well:

```no-highlight
$ node type_annotations.js
/Users/sequoia/tmp/type_annotations.js:3
function foo(x: string, y: number): string {
              ^
SyntaxError: Unexpected token :
    at Object.exports.runInThisContext (vm.js:76:16)
    at Module._compile (module.js:542:28)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3
```

# Expected

Dependency free function in a **JavaScript** file (`.js` extension) runs in a **JavaScript** interpreter.

# Actual

Code in a **JavaScript** file (`.js` extension) file throws a syntax error when run by a javascript 
interpreter.

# Problem

The `.js` extension, particularly in web applications, is reserved for **JavaScript** type files. See: [What is the JS file extension](https://www.reviversoft.com/file-extensions/js). Adding non-JavaScript syntax into `.js` files causes confusion and difficulty for developers (like myself), because the `.js` extension indicates that a file contains **JavaScript** and should generally be interpretable by a **JavaScript** interpreter, possibly with the aid of a tool such as babel to accommodate new **JavaScript** features or proposed features on interpreters with those features not yet implemented.

Separate, non-**JavaScript**, non-proposed-feature syntax should not be in files with the `.js` extension as it will not run in JavaScript interpreters. For example, it would be inappropriate to put the following code into a `.js` file:

```c
#include <stdio.h>
int main()
{
   // printf() displays the string inside quotation
   printf("Hello, World!");
   return 0;
}
```

This is code from another language and will not run in a **JavaScript** interpreter. This is the identical problem with the type annotations found in `type_annotations.js` above.

# Possible solution

Use a different file extension. I suggest `.fjs`.

* Current: `type_annotations.js`
* Suggested Alternative: `type_annotatins.fjs`

The different file extension will alert developers (and IDEs) to the fact that these files are not **JavaScript** files.

**Comment by vkurchatkin** [comment]
You are free to use whatever extension you like

**Comment by gabelevi** [comment]
Thanks for putting so much effort into this issue! It's always a pleasure reading well reported issues like this one!

First, I'd point out that the JavaScript language is being iterated on quite quickly. If you expect your .js files to run on most JavaScript interpreters, you can't use many modern JavaScript features. However the lowest common denominator is not very pleasant to write, so people often use a transpiler like Babel or [`flow-node`](https://github.com/flowtype/flow-remove-types#use-flow-node).

But as @vkurchatkin points you, you are welcome to use other extensions, if you prefer. You can even put code in `.js` files and flow types in `.js.flow` files. Then Flow will ignore the `.js` files and only read the `.js.flow` files.

That said, I'd recommend just using `.js` and transforming your code before running it.

**Comment by Sequoia** [comment]
I don't personally use flow at the moment. My issue is the result of attempting to work on https://github.com/devtools-html/debugger.html and running into unfamiliar syntax in a JS file. I know JS evolves quickly so I googled to find out if there was a proposal for type annotations but did not find one [here](https://github.com/tc39/proposals). JS goes so fast I sometimes am not even aware of new features, so I went to twitter for a sanity check: "Am I crazy? Is this JavaScript??"  It's there I was informed this was flow & not JavaScript.

The project I was working on also uses webpack & babel for all kinds of transforms & things so I had spent some time wading through configs already. It was frustrating to realize not only is the code in a JS file not current JS, neither is it proposed JS, it's just some extra thing from a third party superlang. I like types, I like typescript, and I am not against innovation, but having to put on my deerstalker cap and play Sherlock Holmes to figure out where the **** some syntax came from in a "JavaScript" codebase is frustrating. If Flow were the only library out there that would ever changed or extend JS syntax it wouldn't be such an issue, but there are many.

The TypeScript project uses a separate extension: `.ts`, to indicate that sourcecode with typescript type annotations is not "javascript" source code, even though a `.ts` file can be 100% valid javascript (some typescript features will even work if run through babel as some of the syntax is the same as "es.next"). The JSX project has a `.jsx` extension to tip users off to the fact a file is not "javascript" but "javascript and some other stuff." This allows me as the developer to *immediately* be aware of the fact that this is not vanilla JS but includes some proprietary syntax.

It sounds like there is an "option" to put types alongside javascript (much like `.d.ts` files in the typescript world), so I suppose this is a now a user choice whether to write non-js code into JavaScript files. I will take it up with the project author in this case!

I know I'm anachronistic, behind the times, and just plain old fashioned, but I wish we as a community could agree at least *roughly* on what constitutes "javascript." Oh well 🙃
